include/intercept.h

```diff
@@ -0,0 +1,93 @@
+#ifndef _INTERCEPT_H
+#define _INTERCEPT_H
+
+#include <assert.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdarg.h>
+
+// vDSO wrapper needed includes
+#include <time.h>
+
+#include "cheriintrin.h"
+
+#include "mem_mng.h"
+
+// Forward declarations
+struct Compartment;
+extern struct Compartment* loaded_comp;
+int64_t exec_comp(struct Compartment*, char*, char**);
+struct Compartment* manager_get_compartment_by_id(size_t);
+
+extern void* __capability manager_ddc;
+
+// Number of capabilities required to perform a transition
+#define COMP_RETURN_CAPS_COUNT 2
+
+// Capabilities required to transition back into the manager once compartment
+// execution has finished
+extern void* __capability comp_return_caps[COMP_RETURN_CAPS_COUNT];
+
+// Capability used to point to pair of capabilities when transitioning out of a
+// compartment via an intercept
+extern void* __capability sealed_redirect_cap;
+
+/* Data required to perform the transition for an intercepted function
+ */
+struct func_intercept {
+    char* func_name;
+    uintptr_t redirect_func;
+    void* __capability intercept_pcc;
+};
+
+/* This function expects the argument be passed in `x10`, rather than `x0`, as
+ * well as using `c29` as an argument for the DDC to transition to in order to
+ * allow the intercept to work. It is expected to be called only in very
+ * specific circumstances, and the signature is more illustrative than
+ * functional. As such, it shouldn't be called from a C context, as that will
+ * most likely break things.
+ */
+void intercept_wrapper(void* to_call_fn);
+
+void setup_intercepts();
+
+time_t manager_time(time_t*);
+FILE* manager_fopen(const char*, const char*);
+size_t manager_fread(void* __restrict, size_t, size_t, FILE* __restrict);
+size_t manager_fwrite(void* __restrict, size_t, size_t, FILE* __restrict);
+int manager_fclose(FILE*);
+int manager_getc(FILE*);
+int manager_fputc(int, FILE*);
+int manager___srget(FILE*);
+
+void* my_realloc(void*, size_t);
+void* my_malloc(size_t);
+void my_free(void*);
+int my_fprintf(FILE*, const char*, ...);
+
+size_t my_call_comp(size_t, char*, void*, size_t);
+static const struct func_intercept to_intercept_funcs[] = {
+    /* vDSO funcs */
+    { "time"     , (uintptr_t) manager_time    },
+    /* Mem funcs */
+    { "malloc"   , (uintptr_t) my_malloc       },
+    { "realloc"  , (uintptr_t) my_realloc      },
+    { "free"     , (uintptr_t) my_free         },
+    //{ "fprintf"  , (uintptr_t) my_fprintf      },
```

Comment:
Are these commented for a reason? Should we remove from the PR?

Reply:
I expect we'll be removing the associated intercepted functions anyway, as part of the full move to so compartments, but I can remove these for now, in a08a023e569ec20df0e956eaf941a77a648045d3.